Implement virtio-fs DAX in cloud-hypervisor backend

liubogithub opened the pull request:
This attempts to let clh's `vhost_user_fs` binary work with guest's dax mode so that we can get the benefits of using dax mode.

Issue: #481 

Signed-off-by: Liu Bo <bo.liu@linux.alibaba.com>

sboeuf commented:
@liubogithub thanks for the PR, I'll review this tomorrow :)

sboeuf commented on vhost_rs/src/vhost_user/slave_req_handler.rs:
```diff
@@ -47,6 +49,9 @@ pub trait VhostUserSlaveReqHandler {
     ) -> Result<Vec<u8>>;
     fn set_config(&mut self, offset: u32, buf: &[u8], flags: VhostUserConfigFlags) -> Result<()>;
+    fn set_slave_req_fd(&mut self, _vu_req: SlaveFsCacheReq) {
+        // implicitly do nothing for default implementation.
```
You can remove this comment.

sboeuf commented on vhost_user_fs/src/server.rs:
```diff
@@ -126,6 +134,95 @@ impl<F: FileSystem + Sync> Server<F> {
+            len,
+            flags,
+            moffset,
+            vu_req.unwrap(),
```
Don't `unwrap()` an option like this, please try if it is `Some()`, using the following kind of syntax:
```
if let Some(vu) = vu_req {
...
}
```

sboeuf commented on vhost_user_fs/src/server.rs:
```diff
@@ -63,7 +64,12 @@ impl<F: FileSystem + Sync> Server<F> {
+        &self,
+        mut r: Reader,
+        w: Writer,
+        vu_req: Option<&mut SlaveFsCacheReq>,
```
As mentioned [here](https://github.com/cloud-hypervisor/cloud-hypervisor/issues/481#issuecomment-575226292), I think it'd be cleaner to define a trait to make things more generic. We don't really want the `server.rs` to know that it is dealing with `vhost-user` specifically.

liubogithub replied:
@sboeuf In order to avoid the cyclic dependency, I did the suggested trait inside vhost_rs crate instead of vhost_user_fs,  otherwise,

```
error: cyclic package dependency: package `vhost_rs v0.1.0 (/home/b.liu/1_source/3_alibaba/cloud-hypervisor/vhost_rs)` depends on itself. Cycle:
package `vhost_rs v0.1.0 (/home/b.liu/1_source/3_alibaba/cloud-hypervisor/vhost_rs)`
    ... which is depended on by `vm-virtio v0.1.0 (/home/b.liu/1_source/3_alibaba/cloud-hypervisor/vm-virtio)`
    ... which is depended on by `vhost_user_fs v0.1.0 (/home/b.liu/1_source/3_alibaba/cloud-hypervisor/vhost_user_fs)`
    ... which is depended on by `vhost_rs v0.1.0 (/home/b.liu/1_source/3_alibaba/cloud-hypervisor/vhost_rs)`
    ... which is depended on by `cloud-hypervisor v0.4.0 (/home/b.liu/1_source/3_alibaba/cloud-hypervisor)`
```

sboeuf commented:
@jiangliu it'd be nice if you could have a look to also help provide some guidance here.

liubogithub commented:
> @liubogithub thank you very much for this PR, it looks pretty good :)
> Additionally to the comments I already left, could you pelase:
> 
> * Split the commit `Vhost_user_fs: enable dax mode for vhost_user_fs binary` into two separate commits. First commit would introduce `setupmapping`/`removemapping` and second commit would add support for `set_slave_req_fd()`.
> * Be a bit more detailed in your commit messages, and elaborate about what is being done through each of them, emphasizing on the end goal.
> * Extend the `virtio-fs` integration tests so that we can test the DAX support when using the vhost-user-fs Rust daemon.

Thank you so much for the comments.

I'll update & fix them in v2.

liubogithub commented on vhost_user_fs/src/passthrough.rs:
```diff
@@ -948,7 +948,7 @@ impl FileSystem for PassthroughFs {
         (*vu_req)
             .fs_slave_map(&msg, file.as_raw_fd())
-            .or(Err(io::Error::from_raw_os_error(libc::EINVAL)))
+            .or_else(|_| Err(io::Error::from_raw_os_error(libc::EINVAL)))
```
Yes, one of the CI robots shows "warnings as errors" kind of reports

liubogithub commented on vhost_rs/src/vhost_user/master_req_handler.rs:
```diff
@@ -11,33 +11,37 @@ use std::sync::{Arc, Mutex};
-    fn handle_config_change(&mut self) -> HandlerResult<()> {
-        Err(std::io::Error::from_raw_os_error(libc::ENOSYS))
+    fn handle_config_change(&mut self) -> Result<()> {
+        error_code(std::io::Error::from_raw_os_error(libc::ENOSYS))
```
I made the above error conversion because `SlaveFsCacheReq` needs to errors from Endpoint and I prefer not to losing that error information by mapping it to raw os errors.
(I'll also clarify it in the commit message).

sboeuf commented on vhost_rs/src/vhost_user/slave_fs_cache.rs:
```diff
@@ -9,6 +9,16 @@ use std::os::unix::io::RawFd;
+
+    /// Removemapping operation
+    fn unmap(&mut self, fs: &VhostUserFSSlaveMsg) -> Result<()>;
+}
```
I don't think you should declare the trait here. Having the trait in `vhost_user_fs/src/fs_cache_req_handler.rs` is enough, and you can provide the implementation along with it directly in `vhost_user_fs` crate.

liubogithub replied:
I did it that way but then encountered the cyclic dependency problem (as I mentioned above [here](https://github.com/cloud-hypervisor/cloud-hypervisor/pull/698#discussion_r376673012) ).
Let me check if I can do a workaround.

sboeuf commented:
@jiangliu @jynnantonix 
It'd be nice if you could have a look at this PR.

@jiangliu mostly I'd like your feedback on the modification related to `vhost_rs` crate
@jynnantonix I'd like your feedback regarding the `passthrough.rs` and `server.rs` modifications

liubogithub commented on vhost_user_fs/Cargo.toml:
```diff
@@ -10,3 +10,7 @@ libc = "0.2.65"
 vm-memory = { git = "https://github.com/rust-vmm/vm-memory" }
 vm-virtio = { path = "../vm-virtio" }
+
+[dependencies.vhost_rs]
```
Sure, will do it.

liubogithub commented:
Hi @sboeuf, thank you so much for the comments, I finally understand what you've said and now v3   is here.

v2-> v3
- move all virtiofs specific stuff into vhost_user_fs crate
- use map_err() instead of the error_code() wrapper
- Update Cargo.lock to include the new dependency between vhost_user_fs and vhost_rs

jynnantonix commented on vhost_user_fs/src/filesystem.rs:
```diff
@@ -1043,6 +1045,27 @@ pub trait FileSystem {
+        _ctx: Context,
+        requests: Vec<RemovemappingOne>,
+        vu_req: &mut T,
+    ) -> io::Result<()>;
```
All the other methods in this trait provide default implementations (most just return ENOSYS) so that types that implement this trait are not required to implement those methods.  Not providing a default implementation here means that every type needs to provide an implementation of this method even if they don't support it.

I'd recommend providing the default ENOSYS implementation for both of these as well:

```rust
fn removemapping<T: FsCacheReqHandler>(
    &self,
    _ctx: Context,
    requests: Vec<RemovemappingOne>,
     vu_req: &mut T,
) -> io::Result<()> {
    Err(io::Error::from_raw_os_error(libc::ENOSYS))
}
```

liubogithub replied:
Sure, will update it.

jynnantonix commented on vhost_user_fs/src/fs_cache_req_handler.rs:
```diff
@@ -0,0 +1,23 @@
+use vhost_rs::vhost_user::{Result as VhostUserResult, SlaveFsCacheReq, VhostUserMasterReqHandler};
+
+/// Trait for virtio-fs cache requests operations.  This is mainly used to hide
+/// vhost-user details from virtio-fs's fuse part.
```
The documentation says that the purpose of this trait is to hide vhost-user details from the fuse part but both methods of this trait take a `VhostUserFSSlaveMsg` and return `VhostUserResult` so it's not clear to me what details are actually being hidden.  Instead I would have expected the trait methods to take generic arguments (like the fd offset, cache offset, length, and flags) and then have the implementation of the method construct the `VhostUserFSSlaveMsg` out of them and then convert the `VhostUserResult` into an `io::Result`.

liubogithub replied:
Make sense, I'm going to update it per your suggestion.

jynnantonix commented on vhost_user_fs/src/passthrough.rs:
```diff
@@ -911,6 +913,62 @@ impl FileSystem for PassthroughFs {
+        let file = self.open_inode(inode, open_flags as i32)?;
+        (*vu_req)
+            .map(&msg, file.as_raw_fd())
+            .or_else(|_| Err(io::Error::from_raw_os_error(libc::EINVAL)))
```
I feel like this error should be either wrapped and returned or logged somewhere.  Silently dropping it and replacing it with EINVAL makes it much harder to debug problems in the future.

liubogithub replied:
OK, I've refactored map to return io::Result<()> directly.

jynnantonix commented on vhost_user_fs/src/passthrough.rs:
```diff
@@ -911,6 +913,62 @@ impl FileSystem for PassthroughFs {
+        let mut msg: VhostUserFSSlaveMsg = Default::default();
+
+        for (ind, req) in requests.iter().enumerate() {
+            msg.len[ind] = req.len;
```
Should probably check to make sure that `requests.len() < VHOST_USER_FS_SLAVE_ENTRIES`.  Otherwise this will panic and abort the whole process.

liubogithub replied:
Thanks for pointing this out, it does cause problem because requests may have more items than `VHOST_USER_FS_SLAVE_ENTRIES`, I'm refactoring the code a little bit.

liubogithub commented:
Hi @jynnantonix, thanks for your comments, please take a look at the latest pushed version (v4) here.

v3 -> v4:
- Dropped the patch which changes `VhostUserMasterReqHandler` to return `Result<()>` instead of `HandlerResult<()>`.
- Made map/unmap take generic arguments instead of `VhostUserFSSlaveMsg`.
- Fixed a bug in `unmap` where requests may be sent to master multiple times when requests have more `RemovemappingOne` items than `Vhost_User_Fs_Slave_Entries`.
- Rebased onto the lastest `master` branch.

jynnantonix commented on vhost_user_fs/src/fs_cache_req_handler.rs:
```diff
@@ -0,0 +1,69 @@
+            if i == VHOST_USER_FS_SLAVE_ENTRIES - 1 {
+                self.fs_slave_unmap(&msg)?;
+                msg = Default::default();
+            }
```
This feels a bit clunky.  How about using the `chunks` method from the slice type?

```rust
for chunk in requests.chunks(VHOST_USER_FS_SLAVE_ENTRIES) {
    let mut msg = Default::default();
    for (ind, req) in chunk.into_iter().enumerate() {
        msg.len[ind] = req.len;
        msg.cache_offset[ind] = req.moffset;
    }

    self.fs_slave_unmap(&msg)?;
}
```

liubogithub replied:
Cool, I like it.

jynnantonix commented on vhost_user_fs/src/fs_cache_req_handler.rs:
```diff
@@ -0,0 +1,69 @@
+        flags: u64,
+        fd: RawFd,
+    ) -> io::Result<()> {
+        let mut msg: VhostUserFSSlaveMsg = Default::default();
```
Unrelated to the actual change but is this type defined in the vhost-user protocol somewhere?  It appears to be defined like this:

```rust
pub struct VhostUserFSSlaveMsg {
    /// TODO:
    pub fd_offset: [u64; VHOST_USER_FS_SLAVE_ENTRIES],
    /// TODO:
    pub cache_offset: [u64; VHOST_USER_FS_SLAVE_ENTRIES],
    /// Size of region to map.
    pub len: [u64; VHOST_USER_FS_SLAVE_ENTRIES],
    /// Flags for the mmap operation
    pub flags: [VhostUserFSSlaveMsgFlags; VHOST_USER_FS_SLAVE_ENTRIES],
}
```

which seems kind of backwards to me.  Usually the pattern I've seen is an array of structs rather than a struct with a bunch of arrays.

So I would have expected this to be something like:

```rust
pub struct VhostUserFSSlaveMsg {
    pub fd_offset: u64,
    pub cache_offset: u64,
    pub len: u64,
    pub flags: VhostUserFSSlaveMsgFlags,
}
```

And then the message could be dynamically sized based on the `size` field of the `VhostUserMsgHeader` (up to a max of `VHOST_USER_FS_SLAVE_ENTRIES`).  Something like `[VhostUserFSSlaveMsg; VHOST_USER_FS_SLAVE_ENTRIES]`.

If this hasn't already been finalized in the protocol, I'd like to see if this can be updated.

liubogithub replied:
I think it was to be aligned with `VhostUserFSSlaveMsg` in qemu.

I'm open to the mentioned change, but would like to have it done in a separate patch where  all the `VhostUserFSSlaveMsg` callers would be updated.

jynnantonix replied:
Sure, I wasn't saying that you need to change this here.  I was just asking if this is something that's even possible to change or if it has already made it into the protocol.

sboeuf commented:
@liubogithub what's the status? Have you addressed all comments from @jynnantonix ?

liubogithub commented:
> @liubogithub what's the status? Have you addressed all comments from @jynnantonix ?

Just one more minor fix (using `chunk` to simplify the unmap code), I'll update it soon.

sboeuf commented:
@jiangliu any chance you can have a quick look at it?

sboeuf commented:
@rbradford PTAL

rbradford commented on vhost_user_fs/src/filesystem.rs:
```diff
@@ -1043,6 +1045,31 @@ pub trait FileSystem {
 
+    /// Setup a mapping so that guest can access files in DAX style.
+    #[allow(clippy::too_many_arguments)]
+    fn setupmapping<T: FsCacheReqHandler>(
```
s/setupmapping/setup_mapping/

rbradford replied:
Or could it just be 'map' ?

liubogithub replied:
I'm not against renaming it, but what I did was to follow how CLH has converted virtiofsd's operation names, e.g.
-  do_releasedir -> fn releasedir(), 
- do_fsyncdir -> fn fsyncdir()

Do you think we need to make an exception here?

rbradford replied:
If it's consistent with the rest of the file then I guess that's ok.

rbradford commented on vhost_user_fs/src/filesystem.rs:
```diff
@@ -1043,6 +1045,31 @@ pub trait FileSystem {
+        Err(io::Error::from_raw_os_error(libc::ENOSYS))
+    }
+
+    fn removemapping<T: FsCacheReqHandler>(
```
s/removemapping/remove_mapping/

liubogithub replied:
Ditto.

rbradford commented on vhost_user_fs/src/fs_cache_req_handler.rs:
```diff
@@ -0,0 +1,61 @@
+/// Trait for virtio-fs cache requests operations.  This is mainly used to hide
+/// vhost-user details from virtio-fs's fuse part.
+pub trait FsCacheReqHandler: Send + Sync + 'static {
+    /// Setupmapping operation
```
I don't understand this comment

liubogithub replied:
I'm changing it to
`Setup a dedicated mapping so that guest can access file data in DAX style.`

Does it sound good?

rbradford commented on vhost_user_fs/src/fs_cache_req_handler.rs:
```diff
@@ -0,0 +1,61 @@
+        fd: RawFd,
+    ) -> io::Result<()>;
+
+    /// Removemapping operation
```
Or this one.

liubogithub replied:
`Remove those mappings that provide the access to file data.`
--
how about this?

rbradford commented on vhost_user_fs/src/fuse.rs:
```diff
@@ -513,6 +513,9 @@ pub enum Opcode {
     Rename2 = 45,
     Lseek = 46,
+    CopyFileRange = 47,
+    Setupmapping = 48,
```
s/Setupmapping/SetupMapping/

liubogithub replied:
OK

rbradford commented on vhost_user_fs/src/fuse.rs:
```diff
@@ -513,6 +513,9 @@ pub enum Opcode {
     Lseek = 46,
+    CopyFileRange = 47,
+    Setupmapping = 48,
+    Removemapping = 49,
```
s/Removemapping/RemoveMapping/

liubogithub replied:
OK

rbradford commented on vhost_rs/src/vhost_user/mod.rs:
```diff
@@ -41,6 +41,9 @@ pub use self::slave::SlaveListener;
 mod slave_req_handler;
```
nit: fix the capitalisation in your commit messages (don't capitalise the first letter of your first line of your commit message.)

liubogithub replied:
OK.
